include/ginkgo/core/preconditioner/ilu.hpp

```diff
@@ -0,0 +1,377 @@
+     *
+     * @internal  The second template parameter (which uses SFINAE) must match
+     *            the default value of the general case in order to be accepted
+     *            as a specialization.
+     */
+    template <typename SolverType>
+    struct has_with_criteria<
+        SolverType,
+        xstd::void_t<decltype(std::declval<factory_type_t<SolverType>>()
+                                  .with_criteria(with_criteria_param_type()))>>
+        : std::true_type {};
+
+
+    /**
+     * Generates a default solver of type SolverType.
+     *
+     * Also checks wheather SolverType can be assigned a criteria, and if it
+     * can, it is assigned default values which should be well suited for a
+     * preconditioner.
+     */
+    template <typename SolverType>
+    static xstd::enable_if_t<has_with_criteria<SolverType>::value,
+                             std::unique_ptr<SolverType>>
+    generate_default_solver(const std::shared_ptr<const Executor> &exec,
+                            const std::shared_ptr<const LinOp> &mtx)
+    {
+        constexpr ValueType default_reduce_precision{1e-4};
+        const unsigned int default_max_iters{
+            static_cast<unsigned int>(mtx->get_size()[0])};
+
+        return SolverType::build()
+            .with_criteria(gko::stop::Iteration::build()
+                               .with_max_iters(default_max_iters)
+                               .on(exec),
+                           gko::stop::ResidualNormReduction<>::build()
+                               .with_reduction_factor(default_reduce_precision)
+                               .on(exec))
+            .on(exec)
+            ->generate(mtx);
+    }
+
+    /**
+     * @copydoc generate_default_solver
+     */
+    template <typename SolverType>
+    static xstd::enable_if_t<!has_with_criteria<SolverType>::value,
+                             std::unique_ptr<SolverType>>
+    generate_default_solver(const std::shared_ptr<const Executor> &exec,
+                            const std::shared_ptr<const LinOp> &mtx)
+    {
+        return SolverType::build().on(exec)->generate(mtx);
+    }
+
+private:
+    std::shared_ptr<const LinOp> l_factor_{};
+    std::shared_ptr<const LinOp> u_factor_{};
+    /**
+     * Manages a vector as a cache, so there is no need to allocate one every
+     * time an intermediate vector is required.
+     * Copying an instance will only yield an empty object since copying the
+     * cached vector would not make sense.
+     *
+     * @internal  The struct is present so the whole class can be copyable
+     *            (could also be done with writing `operator=` and copy
+     *            constructor of the enclosing class by hand)
+     */
+    mutable struct cache_struct {
+        cache_struct() = default;
+        ~cache_struct() = default;
+        cache_struct(const cache_struct &other) {}
+        cache_struct &operator=(const cache_struct &) { return *this; }
```

Comment:
I think the move construct is missing. Sonarqube will probably complain that you are not following rule of 5/0

Reply:
Thanks, fixed it.